cmd/builder/internal/builder/main.go

```diff
@@ -131,29 +140,66 @@ func GetModules(cfg Config) error {
 	}
 
 	// ambiguous import: found package cloud.google.com/go/compute/metadata in multiple modules
-	if err := runGoCommand(cfg, "get", "cloud.google.com/go"); err != nil {
+	if _, err := runGoCommand(cfg, "get", "cloud.google.com/go"); err != nil {
 		return fmt.Errorf("failed to go get: %w", err)
 	}
 
-	if err := runGoCommand(cfg, "mod", "tidy", "-compat=1.21"); err != nil {
+	if _, err := runGoCommand(cfg, "mod", "tidy", "-compat=1.21"); err != nil {
 		return fmt.Errorf("failed to update go.mod: %w", err)
 	}
 
+	if cfg.SkipStrictVersioning {
+		return downloadModules(cfg)
+	}
+
+	// Perform strict version checking.  For each component listed and the
+	// otelcol core dependency, check that the enclosing go module matches.
+	modulePath, dependencyVersions, err := cfg.readGoModFile()
+	if err != nil {
+		return err
+	}
+
+	corePath, coreVersion := cfg.coreModuleAndVersion()
+	if dependencyVersions[corePath] != coreVersion {
```

Comment:
The error here and below would be a bit confusing if `corePath` is not on `dependencyVersions`. While this would be a weird situation that I don't think we can encounter now, IMO for future-proofing we should do the full check  (`corePathVersion, ok := dependencyVersions[corePath]`) and provide a more appropriate error message if `ok` is false